### PR TITLE
NPC "completion" / MissionAction conversations can kill the player and/or NPC

### PR DESCRIPTION
**Pull request by tehhowch**
Refs #3206 , #3129, #3228 

 - Adds a new player-death keyword, `explode`, to resolve the ambiguity over how the player dies while in space
   - If the player's ship should be destroyed, use `explode`. The flagship will explode.
   - If the player's ship should be captured, use `die`. The player dies, and his ship is taken over by the ship it was boarding. (If `die` is used and there is no boarding ship, the flagship changes government, simulating a mutiny.)
 - If you're dying as a result of a conversation, you aren't going to be shown the BoardingPanel
 - `launch`, `depart`, and `flee` kill the NPC associated with a conversation, even if not boarded (say, if your outfit scan activated some embedded explosives' trigger mechanism...), unless you're dying too.

### CODE REVIEW

**Comment by tehhowch** [comment]
Examples:
```
	npc board
		government "Pirate"
		personality target entering unconstrained harvests plunders
		ship "Protector (Proton)" "Bastard of the Pond"
		conversation
			`"Good job, <first>! I like this ship!"`
			`	(You are killed.)`
				die
```
```
	on offer
		conversation
			`Your ship's dosimetry sensors begin crackling as you dock with <ship>, but before you can abort, a sharp clanking echoes from your hull. The momentary silence is followed by a chilling, robotic voice emanating from your ship's P.A. system:`
			``
			`	"Radiological threat detected."`
			``
			`A highly radioactive, shaped-charge explosive has attached to your hull. Uh oh.`
			`	(You disintegrate.)`
				explode
```

**Comment by tehhowch** [comment]
Added support for #3228 as well

**Comment by Pointedstick** [comment]
Wow, amazing! Almost there for my mutiny mission sketched out in https://github.com/endless-sky/endless-sky/issues/3228#issuecomment-343784930, but not quite... In the example mission above, I don't want the ship to blow up. The mercenaries capture it and fly off. Maybe `die` without `explode` could do that? In other words, just undo or don't implement the condition "(If die is used and there is no boarding ship, the flagship explodes.)"

**Comment by tehhowch** [comment]
Right now that "explode if dead but not boarded" behavior is basically a guard against referencing a ship that doesn't exist. I haven't really checked what else can be done - it's possible just changing the "is yours" property would be sufficient.

**Comment by tehhowch** [comment]
@Pointedstick should be all set now :+1: 
For now I've opted to have the mutineer government be the "Independent" government. A future PR could extend how this is selected.

(I also added some reparenting so that, should the player's flagship be captured in any manner, the rest of the player's fleet does not take orders from the captured ship. This means they won't follow it around, wait for it to jump, etc)

**Comment by Pointedstick** [comment]
Outstanding. I have some fun missions queued up to submit once this is in...

**Comment by tehhowch** [comment]
Confirmed these all still work after the change in ShipEvent handling :+1:

File: source/PlayerInfo.cpp [hunk 1]
```diff
@@ -2579,3 +2621,46 @@ bool PlayerInfo::CanBeSaved() const
...
+			flagship->Destroy();
+		else
+		{
+			flagship->SetIsYours(false);
+			// If the flagship is boarding an NPC, the player's death results in its
+			// capture. Otherwise, consider the player to have died due to a mutiny.
+			// Change the flagship government to remove it from the EscortDisplay HUD.
+			if(boardingShip)
+			{
+				flagship->SetSwizzle(boardingShip->GetSwizzle());
```
Comment (endless-sky) [inline source/PlayerInfo.cpp]: Ships don't actually change swizzle when they're captured, since that requires a new paint job. (They don't change until you land on a planet.)

File: source/PlayerInfo.cpp [hunk 2]
```diff
@@ -1634,6 +1645,37 @@ void PlayerInfo::HandleEvent(const ShipEvent &event, UI *ui)
...
+		// Nothing to do.
+	}
+	else if(response == Conversation::DECLINE || response == Conversation::FLEE)
+	{
+		// Nothing to do.
+	}
+	else if(response == Conversation::DEFER || response == Conversation::DEPART)
+	{
+		// Nothing to do.
+	}
```
Comment (endless-sky) [inline source/PlayerInfo.cpp]: I don't think all the above conditions need to be here. To my mind it would be clearer to only have an `if` for the responses that actually have a result.

File: source/Ship.cpp [hunk 3]
```diff
@@ -1980,8 +1982,8 @@ void Ship::WasCaptured(const shared_ptr<Ship> &capturer)
 	// Transfer some crew over. Only transfer the bare minimum unless even that
 	// is not possible, in which case, share evenly.
 	int totalRequired = capturer->RequiredCrew() + RequiredCrew();
-	int transfer = RequiredCrew();
-	if(transfer)
+	int transfer = RequiredCrew() - crew;
+	if(transfer > 0)
 	{
 		if(totalRequired > capturer->Crew())
 			transfer = max(1, (capturer->Crew() * RequiredCrew()) / totalRequired);
```
Comment (endless-sky) [inline source/Ship.cpp]: If you're adding support for capturing a ship that still has crew, I think this would need to be:

```C++
if(totalRequired > capturer->Crew() + crew)
    transfer = max(crew ? 0 : 1, (capturer->Crew() * transfer) / totalRequired);
```

That is, if there's not enough crew to go around, split the crew from both ships evenly. And, if the victim has no crew, make sure to give it at least one.

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -2579,3 +2621,46 @@ bool PlayerInfo::CanBeSaved() const
...
+		Die(true);
+	else
+	{
+		Die();
+		auto it = find(ships.begin(), ships.end(), flagship);
+		if(response == Conversation::EXPLODE)
+			flagship->Destroy();
+		else
+		{
+			flagship->SetIsYours(false);
```
Comment (endless-sky) [inline source/PlayerInfo.cpp]: Technically speaking, the flagship's crew should be reduced by 1 (since you're dead). But, that would create an issue if you're boarded by someone with only 1 crew, and you only have 1 crew too, because there would not be enough crew to control both ships.

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -2579,3 +2621,46 @@ bool PlayerInfo::CanBeSaved() const
...
+				it = ships.erase(it);
+		}
+		
+		// Let the player's fleet pick a new flagship to follow, if possible.
+		for( ; it != ships.end(); ++it)
+			if((*it)->CanBeFlagship())
+				break;
+		const shared_ptr<Ship> newFlagship = it != ships.end() ? *it : nullptr;
+		for(const shared_ptr<Ship> &escort : ships)
+			escort->SetParent(escort != newFlagship ? newFlagship : nullptr);
```
Comment (endless-sky) [inline source/PlayerInfo.cpp]: This seems unnecessary because the game is already over - it's no big deal if your escorts scatter instead of picking a new leader and departing together.

File: source/ConversationPanel.cpp [hunk 6]
```diff
@@ -343,21 +343,22 @@ void ConversationPanel::Goto(int index, int choice)
...
-	// being forced to leave via LAUNCH, FLEE, or DEPART destroys it. If it
-	// is a hostile NPC and not destroyed, and its mission (e.g. "leave me
-	// alone, here's X as payment") was not accepted, open a BoardingPanel
-	// to allow plundering it - unless you're dead.
-	if(node != Conversation::DIE && player.BoardingShip())
+	// If this is a conversation offered from an NPC, ending it via LAUNCH,
+	// FLEE, or DEPART destroys the NPC. For hostile NPCs that are boarded
+	// and not being destroyed, show the BoardingPanel unless the player is
+	// being killed or the conversation is exiting via ACCEPT.
+	if(node != Conversation::DIE && node != Conversation::EXPLODE && ship)
```
Comment (endless-sky) [inline source/ConversationPanel.cpp]: Note: the `ship` pointer could be a ship you just scanned that triggered a conversation, in which case making it explode seems weird.

File: source/PlayerInfo.cpp (hunk 2 above)
Comment (tehhowch) [reply source/PlayerInfo.cpp]: I agree, they don't need to be here. I added them because I thought it was possibly beneficial to explicitly convey that "yes, these responses are possible, here is where they are handled, and they do nothing w.r.t the player."

File: source/PlayerInfo.cpp (hunk 4 above)
Comment (tehhowch) [reply source/PlayerInfo.cpp]: I thought of reducing it, but since the player is dead and can no longer inspect the ship that just boarded them, it was better to keep as-is than reduce it and perhaps cause an issue.

File: source/PlayerInfo.cpp (hunk 5 above)
Comment (tehhowch) [reply source/PlayerInfo.cpp]: I added this so that the player's fleet doesn't simply obey whoever killed the player, and follow that the player's "old" ship around (since it isn't destroyed, they won't scatter).

File: source/Ship.cpp (hunk 3 above)
Comment (tehhowch) [reply source/Ship.cpp]: You are correct

File: source/PlayerInfo.cpp (hunk 1 above)
Comment (tehhowch) [reply source/PlayerInfo.cpp]: Indeed, here it was added solely as a visual indicator of "what happened," even if the mission-writer didn't really expound on the details.

File: source/ConversationPanel.cpp (hunk 6 above)
Comment (tehhowch) [reply source/ConversationPanel.cpp]: `"We've sabotaged this <npc> with explosives, and they're set to go off when you scan their cargo"`

**Comment by endless-sky** [comment]
Can you think of any reason why the original Die() function needed a version that did not clear the player's fleet? It seems like there's no reason for that boolean switch.

If the original Die() always cleared the player's whole fleet, cleared the flagship pointer, and cleared IsYours() on every one of the player's ships for good measure, just calling Die() would be enough to make the player's fleet fly off on its own. Then, instead of your whole extra Die() function, all you'd have to do is destroy the flagship if responding to an EXPLODE response (and not landed, and the flagship exists).

Also, your code goes halfway toward eliminating the need for PlayerInfo::BoardingShip() by passing a shared pointer to the ship in to ConversationPanel. If Mission::CanOffer() and Mission::Instantiate() had an optional ship pointer as a second argument, I think tracking `boardingShip` in PlayerInfo and clearing it at the proper time would become unnecessary.

In other words, I think there would be ways to implement this that would make the code a little bit simpler instead of making it more complex.

File: source/ConversationPanel.cpp (hunk 6 above)
Comment (Pointedstick) [reply source/ConversationPanel.cpp]: > "We've sabotaged this \<npc\> with explosives, and they're set to go off when you scan their cargo"

[gets some ideas]

**Comment by tehhowch** [comment]
I'll have to mull that change over in the next few days (feel free to take the initiative and do it for me if you're feeling inspired 👍). Otherwise, I can probably get to it Monday evening or so.

**Comment by tehhowch** [comment]
Based on the commits involved, the boolean version was added because moving `ships.clear()` into `Die()` (from `MissionCallback()`) "broke the normal death" wherein the flagship is destroyed in flight (presumably because all the remaining player ships simply vanished). PI::HandleEvent (expected) and the fine-handling of `atrocity` while landed (almost certainly a bug, and now addressed in #3208) are currently the only non-`allShipsDie` caller on `master`.
I feel this non-`allShipsDie` behavior is what needs to be replicated when the player dies to capture / mutiny, and that clearing of the player's ship list only needs to be done if the player is on a planet (lest they simply press D to take off).

**Comment by tehhowch** [comment]
With 6231420 (as well as #3208), the player being detained for life due to landing on a planet and getting caught with atrocity-level outfits actually sticks (instead of being able to simply press `D` and take off, or do anything else while landed, from memory.

**Comment by tehhowch** [comment]
@endless-sky The only bits I'm not 100% satisfied with are the boarding checks that guard the player being captured, and showing the BoardingPanel after rejecting the NPC's mission. Other than that, I've successfully eliminated BoardingShip, and eliminated the need for the second callback.

File: source/ConversationPanel.cpp [hunk 7]
```diff
@@ -343,21 +343,27 @@ void ConversationPanel::Goto(int index, int choice)
...
+	// block (e.g. scanning or boarding or killing all required targets).
+	if(node == Conversation::DIE || node == Conversation::EXPLODE)
+		player.Die(node, ship);
+	else if(ship)
 	{
+		// A forced-launch ending (LAUNCH, FLEE, or DEPART) destroys any NPC.
 		if(Conversation::RequiresLaunch(node))
-			player.BoardingShip()->Destroy();
-		else if(player.BoardingShip()->GetGovernment()->IsEnemy()
+			// TODO: This may need a ShipEvent, but not all instances
```
Comment (endless-sky) [inline source/ConversationPanel.cpp]: I think this is safe, because Engine::MoveShip() also creates a DESTROY event when a ship has finished exploding, to handle self-destructing ships.

File: source/ConversationPanel.cpp [hunk 8]
```diff
@@ -343,21 +343,27 @@ void ConversationPanel::Goto(int index, int choice)
...
+			// are necessarily caused by the player.
+			ship->Destroy();
+		// Only show the BoardingPanel for a hostile NPC that is being boarded.
+		// (NPC completion conversations can result from non-boarding events.)
+		else if(ship->Position().Distance(player.Flagship()->Position()) <= 1.
```
Comment (endless-sky) [inline source/ConversationPanel.cpp]: Could this check `&ship == player.BoardingShip().get()` instead? Or is there a reason that wouldn't work? Relying on the distance to identify a ship you're boarding just seems like a bit of a hack.

File: source/ConversationPanel.h [hunk 9]
```diff
@@ -124,6 +125,10 @@ template <class T>
 	// If specified, this is a star system to display with a special big pointer
 	// when the player brings up the map. (Typically a mission destination.)
 	const System *system;
+	// If specified, this is a pointer to the ship that this conversation
+	// acts upon (e.g. the ship failing a "flight check", or the NPC you
+	// have boarded).
+	std::shared_ptr<Ship> ship = nullptr;
```
Comment (endless-sky) [inline source/ConversationPanel.h]: shared_ptr<> doesn't need to be initialized, because its default constructor will set it to null.

File: source/ConversationPanel.cpp (hunk 8 above)
Comment (endless-sky) [reply source/ConversationPanel.cpp]: Oops, never mind, now I remember that the point was to eliminate PlayerInfo::BoardingShip().

Comment (endless-sky) [reply source/ConversationPanel.cpp]: I wonder if checking `ship->IsYours()` would work. The only times a ConversationPanel has a ship is if 1. it's a flight check message (for one of your ships) or 2. it's a conversation being shown because you boarded a ship. So, if the ship is not yours, I think it must be a ship you're boarding.

Comment (tehhowch) [reply source/ConversationPanel.cpp]: NPC completion also packages a ship into ConversationPanel (hence the code comments), and the ship isn't necessarily being boarded - it might have been scanned, destroyed, disabled, etc.
